### PR TITLE
[Contracts/Deprecation] Provide a generic function and convention to trigger deprecation notices

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR results from a discussion we had yesterday with @alcaeus, @beberlei and others at the Symfony UG meetup in Berlin (thanks SensioLabs for bringing me there).

It provides a generic function and convention to trigger deprecation notices, which could be a better alternative to the `@trigger_error(..., E_USER_DEPRECATED)` calls that we use currently.

This proposed package would provide a single global function named `deprecated()`.
Its purpose is to trigger deprecations in a way that can be silenced on production environment
by using the `zend.assertions` ini setting and that can be caught during development to generate reports.

The function requires at least 3 arguments:
 - the name of the package that is triggering the deprecation
 - the version of the package that introduced the deprecation
 - the message of the deprecation
 - more arguments can be provided: they will be inserted in the message using `printf()` formatting

Example:
```php
deprecated('symfony/blockchain', 8.9, 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
```

This will generate the following message:
`Since symfony/blockchain 8.9: Using "bitcoin" is deprecated, use "fabcoin" instead.`

Check #35550 to see how using this function could look like on Symfony itself.

### CODE REVIEW

File: src/Symfony/Contracts/Deprecation/deprecated.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+/**
+ * Triggers a deprecation.
+ *
+ * As recommended for prod, turn the "zend.assertions" ini setting to 0 or -1 to disable deprecation notices.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+function deprecated($package, $version, $message)
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/Deprecation/deprecated.php]: The function doesn't use type hints nor variadics on purpose, to make it as fast as possibly achievable.

**Comment by ro0NL** [comment]
is the `@` operator required to meet the proposed purpose?

>  trigger deprecations in a way that can be silenced on production environment
by using the zend.assertions ini setting and that can be caught during development to generate reports

i.e. is it worth clarifying the difference when used y/n?

**Comment by OskarStark** [comment]
In your example it looks like 8.9 is a float but it should be a string.

File: src/Symfony/Contracts/Deprecation/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+The function requires at least 3 arguments:
+ - the name of the package that is triggering the deprecation
+ - the version of the package that introduced the deprecation
+ - the message of the deprecation
+ - more arguments can be provided: they will be inserted in the message using `sprintf()`
+
+Example:
+```php
+@deprecated('symfony/blockchain', 8.9, 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
```
Comment (greg0ire) [inline src/Symfony/Contracts/Deprecation/README.md]: As per @OskarStark 's comment
```suggestion
@deprecated('symfony/blockchain', '8.9', 'Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin');
```

File: src/Symfony/Contracts/Deprecation/deprecated.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Triggers a deprecation.
+ *
+ * As recommended for prod, turn the "zend.assertions" ini setting to 0 or -1 to disable deprecation notices.
+ *
+ * The function doesn't use type hints nor variadics on purpose, to make it as fast as possibly achievable.
```
Comment (greg0ire) [inline src/Symfony/Contracts/Deprecation/deprecated.php]:
```suggestion
 * The function doesn't use type declarations nor variadics on purpose, to make it as fast as possibly achievable.
```

**Comment by OskarStark** [comment]
Let’s say I will mute it via zend.assertions, I would also mute my follow code right?

assert(null !== $object);

If yes, this could lead to unexpected behavior...

**Review by greg0ire — COMMENTED** [review]
I think this is a great move, especially the `assert` part! 👍

**Comment by nicolas-grekas** [comment]
> In your example it looks like 8.9 is a float but it should be a string.

In PHP there is a really nice feature which is called type coercion. It's really nice because we can delegate these kind of details to the engine when they don't matter and let it deal with the concern for us!

**Comment by nicolas-grekas** [comment]
> Let’s say I will mute it via zend.assertions, I would also mute my follow code right? `assert(null !== $object);` If yes, this could lead to unexpected behavior...

That's totally expected as that exactly what assert had been designed for. The opposite would be the unexpected behavior.

**Comment by OskarStark** [comment]
> That's totally expected as that exactly what assert has been designed for. The opposite would be the unexpected behavior.

Let me be more clear on this topic, from a DX perspective. You explain, that one can mute this by setting setting X to Y but it is not clear (unless you check the method body itself), that the "muting" is done by deactivating the `assert()` function and not everybody knows (me too until now) that this ini settings exists.

Apart from this I really welcome this PR and I am 👍

**Comment by nicolas-grekas** [comment]
@OskarStark I'm not sure I understand what you want me to do, can you please tell me how you think I should update the PR?

File: src/Symfony/Contracts/Deprecation/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+Symfony Deprecation Contracts
+=============================
+
+A generic function and convention to trigger deprecation notices.
+
+This package provides a single global function named `deprecated()`.
+Its purpose is to trigger deprecations in a way that can be silenced on production environment
```
[the remark this answers is not in the document]
Comment (stof) [reply src/Symfony/Contracts/Deprecation/README.md]: actually, the first suggest should be `a`, not `an`

File: src/Symfony/Contracts/Deprecation/deprecated.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Triggers a deprecation.
+ *
+ * As recommended for prod, turn the "zend.assertions" ini setting to 0 or -1 to disable deprecation notices.
```
Comment (fancyweb) [inline src/Symfony/Contracts/Deprecation/deprecated.php]: I think we could explain here that turning this setting to 0 actually disable all assertions and since the deprecation is "wrapped" in an assertion, it will also disable deprecations.

File: src/Symfony/Contracts/Deprecation/deprecated.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * The function doesn't use type hints nor variadics on purpose, to make it as fast as possibly achievable.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+function deprecated($package, $version, $message)
+{
+    assert(trigger_error("Since $package $version: ".sprintf($message, ...array_slice(func_get_args(), 3)), E_USER_DEPRECATED));
```
Comment (stof) [inline src/Symfony/Contracts/Deprecation/deprecated.php]: shouldn't it use a silenced deprecation so that it is not turned into an exception by error handlers ?

File: src/Symfony/Contracts/Deprecation/deprecated.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * Triggers a deprecation.
+ *
+ * As recommended for prod, turn the "zend.assertions" ini setting to 0 or -1 to disable deprecation notices.
+ *
+ * The function doesn't use type hints nor variadics on purpose, to make it as fast as possibly achievable.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+function deprecated($package, $version, $message)
```
Comment (stof) [inline src/Symfony/Contracts/Deprecation/deprecated.php]: please add phpdoc describing arguments, as they don't have typehints.

File: src/Symfony/Contracts/Deprecation/README.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+by using the `zend.assertions` ini setting and that can be caught during development to generate reports.
+
```
Comment (OskarStark) [inline src/Symfony/Contracts/Deprecation/README.md]:
```suggestion

**NOTE:** By disabling `zend.assertions` all `assert()` method calls in your code will not work anymore!
```

@nicolas-grekas sth like this.

TLDR:
One think he will JUST silence deprecations but could deactivate important assert()-calls in his code! I want to make the user aware of this! Thats all

**Comment by OskarStark** [comment]
This reads more natural:
```diff
- Deprecated: Since symfony/foo 12.3: lala lala
+ Deprecated since symfony/foo 12.3: lala lala
```

Shall we do some string manipulation or keep the two `:` in the sentence?

File: src/Symfony/Contracts/Deprecation/README.md (hunk 8 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/README.md]: That is way too scary.
People that use `assert()` already know that the only reason why they do so is to benefit from this feature of PHP.
The very promise and the only use case for `assert()` is to disable-able in prod to turn dev-only checks into no-op in prod.
We're now going to teach against this feature.

File: src/Symfony/Contracts/Deprecation/deprecated.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: I think there is zero reasons to do so.
The side that must take care of this is the author, not the users.
If authors use `assert()` they MUST be prepared to it being disabled in prod, that's the ONLY reason why this language construct exists.

**Comment by nicolas-grekas** [comment]
There is no "Deprecated" prefix in my proposal because all the time, it's found in the message.

File: src/Symfony/Contracts/Deprecation/README.md (hunk 8 above)
Comment (alcaeus) [reply src/Symfony/Contracts/Deprecation/README.md]: Can't agree more with @nicolas-grekas: the very purpose of `assert` in PHP is that it has no side-effects whatsoever in production, not even the overhead of the method call to `trigger_error`. That makes it the ideal solution. People that use `assert` for other purposes (e.g. to force strict type checking in production) should be aware of what enabling `zend.assertions` on production environments entails. We should neither advocate that use-case, nor cater to it.

**Review by alcaeus — APPROVED** [review]
I agree with @stof about calling `@trigger_error` by default. We have noticed that using `trigger_error` without the suppression operator is a big no-no (see Symfony 2.7), and thus I think that we should not allow people that trigger deprecations to not make the method silent.

Other than that, I see this package as a very good evolution of the current deprecation logic in Symfony. It also allows people to provide different implementations for `deprecated()` by creating a package that uses `replace` in its composer.json. This gives people that want to log deprecations without the use of the PHP error handler mechanism to provide their own logic.

All in all, big 👍 from me for merging this and making this 1.0 of `symfony/deprecation-contracts` 🎉

**Review by ostrolucky — CHANGES_REQUESTED** [review]
I was present at meetup, but we quickly changed topics so I didn't manage to voice my opinion. I don't agree that deprecations should be tied to zend.assertions. There is plenty people who log deprecations in production, but don't want to make normal assertions causing crashes in production, which is what would happen when these people are forced to enable assertions in order to log deprecations.

**Comment by OskarStark** [comment]
> There is no "Deprecated" prefix in my proposal because all the time, it's found in the message.

This is the exact final output, based on:
https://3v4l.org/CIZch

**Comment by nicolas-grekas** [comment]
PR updated, comments addressed.

> is the @ operator required to meet the proposed purpose?

updated

> This is the exact final output, based on: 3v4l.org/CIZch

I think that doesn't work in real life, with real messages.

> people who log deprecations in production, but don't want to make normal assertions causing crashes in production

Normal assertions can be configured to trigger a warning instead of an exception, and warnings can be turned to logs in prod (by default on Symfony). So I think this PR has a solution for every situations.

**Comment by beberlei** [comment]
@ostrolucky `trigger_error` always returns true in this function, so the assertion is only a mechanism to make the AST assert optimization remove that code completly in production, making it essentially:

```php
function deprecated($component, $version, $message)
{
}
```

**Comment by beberlei** [comment]
@nicolas-grekas hm now that I think about it, the assert woudl only make the deprecations visible in development, iff you have zend.assertion configured. By default its off, so nobody would see the deprecations anymore unless you plan to enable them in the Kernel when debug=true

**Comment by ostrolucky** [comment]
> assertion is only a mechanism to make the AST assert optimization remove that code completly in production

Yes I know but it's not very related. It's certainly nice that unlike before, deprecation triggers are not executed by default in production at all (although it doesn't seem anybody measured impact compared to normal error handler which ignores notices). My point though was that enabling zend assertions will cause problems in other parts of the codebase, as suddenly codebase might produce lot of warnings/errors - other parts than triggering deprecations. Another thing is enabling it will cause much greater negative performance impact than it is now. These points are especially true for packages which started using `assert` as glorified `/** @var Foo $foo */`, like proposed at https://github.com/doctrine/coding-standard/pull/47

I'm also pretty sure this change will lock out some folks from getting deprecation notices. *At least* during development, people certainly want to convert warnings to exceptions. They *also* want to know about deprecations. However as soon as one of their dependencies have failing assertion, they have no control over that and will have to disable assertions, with which they loose deprecation notices as well.

I'm also missing what are the pros of this solution, as PR description explains what was done only, not *why*.

**Comment by alcaeus** [comment]
@ostrolucky quoting Marco from the linked PR:

> Assertions are supposed to:
> 1. Never fail
> 2. Be disabled in production

With that in mind, we should ignore the first use-case as we don't want people to enable assertions in productions, period. That is also why this is an advantageous solution: it has no performance impact on production, unlike the previous solution which always hits the error handler.

As for locking folks out, this could be handled in Symfony by checking whether assertions are in production mode and throwing an exception (as they can't be re-enabled). If the setting is `0` (which only skips them from being called) then Symfony can change the ini setting to `1` which enables deprecation warnings.

**Comment by ostrolucky** [comment]
>    Assertions are supposed to:
>
>        Never fail
>        Be disabled in production

This is why I am making my point. People *don't want to enable assertions in production*. But now they *will have to* if they want to log deprecations there.


> we don't want people to enable assertions in productions

Not wanting people do it and not supporting it are two very different things. It would need official statement from symfony core maintainer to clarify that. You can't be mad when people try to find out what deprecations are triggered in production

> it has no performance impact on production

If you don't enable assertions then no, but even then it has potentially massive performance impact in developemnt, which is already notoriously very slow with Symfony due to container compilations. If there is thousand regular assertions in execution path but from those only single deprecation, enabling assertions for catching this deprecation will cause 1001 extra code/assertions to be executed.

If this micro optimization is important to you, it would be better to just overwrite function with empty one in production, instead of relying on letting assertions to do it implicitly. Perhaps DI container can generate such function. Then we won't have problem with mixed up two different usages of assertions.

**Comment by nicolas-grekas** [comment]
The function definition is now wrapped in a `function_exists('deprecated')` check. That gives one more way to decide for the behavior ppl want. I think this should solve @ostrolucky's concern.

**Comment by nicolas-grekas** [comment]
Actually, we might not need the `function_exists()` check I just added:
a better alternative would be for projects that want another behavior to "replace" the package and provide their implementation instead.

**Comment by nicolas-grekas** [comment]
`function_exists()` now replaced by a hint in the docblock.

**Comment by nicolas-grekas** [comment]
Check #35550 to see how using this function could look like on Symfony itself.

**Review by chalasr — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Contracts/Deprecation/deprecated.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * @param string $message The message of the deprecation; more arguments can be provided:
+ *                        they will be inserted in the message using printf() formatting
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+function deprecated($package, $version, $message)
+{
+    assert(@trigger_error(
+        ($package || $version ? "Since $package $version: " : '')
+        .(3 < func_num_args() ? vprintf($message, array_slice(func_get_args(), 3)) : $message),
```
Comment (lyrixx) [inline src/Symfony/Contracts/Deprecation/deprecated.php]: 1. It's not possible to use variadic here ?
2. Is it intended to use `vprintf` and not `sprintf`, in the PR description, you talk about `sprintf`

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: 1. yes it is, but it wouldn't provide anything - it would even be a bit slower
2. description updated :)

File: src/Symfony/Contracts/Deprecation/deprecated.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: I'm not sure it makes a real difference. 
I already benched this, and the difference is small: https://github.com/lyrixx/php-bench/commit/ea0ff97ad01b91e554377760028f4588beba25d7
And It's not gonna be hit that much.

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: I've already seen the same deprecation triggered in a loop million times.
And there is no benefit to using types here.

File: src/Symfony/Contracts/Deprecation/deprecated.php (hunk 9 above)
Comment (lyrixx) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: I did a [bench](https://github.com/lyrixx/php-bench/commit/a9bed4b5fbefdadcc1632487126f5d9eaf11ca26) about variadic

```
+------------------------------+---------+-------+
| subject                      | mean    | diff  |
+------------------------------+---------+-------+
| benchWithAndEmpty            | 0.062μs | 1.15x |
| benchWithAndEmptyNotEmpty    | 0.110μs | 2.05x |
| benchWithoutAndEmpty         | 0.053μs | 1.00x |
| benchWithoutAndEmptyNotEmpty | 0.113μs | 2.12x |
+------------------------------+---------+-------+
```

It's faster when "$args" is not empty :joy: 
I would prefer variadic version, perso it's easier to read

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: Actually, the code is optimized for prod setup, ie when zend.assertions=0

**Review by lyrixx — APPROVED** [review]

File: src/Symfony/Contracts/Deprecation/deprecated.php (hunk 9 above)
Comment (lyrixx) [inline src/Symfony/Contracts/Deprecation/deprecated.php]: I think you mean `vsprintf`

File: src/Symfony/Contracts/Deprecation/README.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+=============================
+
+A generic function and convention to trigger deprecation notices.
+
+This package provides a single global function named `deprecated()`.
+Its purpose is to trigger deprecations in a way that can be silenced on production environments
+by using the `zend.assertions` ini setting and that can be caught during development to generate reports.
+
+The function requires at least 3 arguments:
+ - the name of the package that is triggering the deprecation
```
Comment (fabpot) [inline src/Symfony/Contracts/Deprecation/README.md]: Should probably mention that we are talking about the Composer package name here to avoid any ambiguity.

File: src/Symfony/Contracts/Deprecation/README.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+ - the version of the package that introduced the deprecation
+ - the message of the deprecation
+ - more arguments can be provided: they will be inserted in the message using `printf()` formatting
```
Comment (fabpot) [inline src/Symfony/Contracts/Deprecation/README.md]: I would remove this possibility as it's not part of the signature.

Using `deprecated('symfony/blockchain', 8.9, sprintf('Using "%s" is deprecated, use "%s" instead.', 'bitcoin', 'fabcoin'));` seems readable enough.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Contracts/Deprecation/README.md (hunk 11 above)
Comment (alcaeus) [reply src/Symfony/Contracts/Deprecation/README.md]: It could be part of the signature by using a variadic argument. Would you be ok with that or would you always prefer people use `sprintf`?

**Comment by fabpot** [comment]
I think that the function name should be a verb, like `trigger_deprecation()` instead of `deprecated()`. That would also be more consistent with built-in PHP functions.

File: src/Symfony/Contracts/Deprecation/README.md (hunk 10 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/README.md]: updated to "the name of the *Composer* package [...]"

File: src/Symfony/Contracts/Deprecation/README.md (hunk 11 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/README.md]: I think using sprintf() is frequent enough to be worth being builtin.
Signature updated with an explicit variadic argument, does this fix your concern?

Comment (fabpot) [reply src/Symfony/Contracts/Deprecation/README.md]: Having an explicit argument works for me.

**Comment by nicolas-grekas** [comment]
> I think that the function name should be a verb, like trigger_deprecation() instead of deprecated().

I'm not sold on this one, because the function is explicitly designed to *not* trigger a deprecation in prod. Any verb would kinda lie to the reader.
In this regard, while this is a function call, this is closer to an annotation to me: it states that this code path is deprecated. What happens when this code path is run is configuration dependent. The caller doesn't need not want to care. Said another way, that's a declarative statement.

**Comment by OskarStark** [comment]
> it states that this code path is deprecated.

This explanation fits very well 👍

File: src/Symfony/Contracts/Deprecation/deprecated.php (hunk 3 above)
Comment (Tobion) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: > The function doesn't use type hints nor variadics on purpose

The statement seems wrong. It does use variadics `...$args`

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Contracts/Deprecation/deprecated.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Deprecation/deprecated.php]: fixed!

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Contracts/Deprecation/README.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+by using the `zend.assertions` ini setting and that can be caught during development to generate reports.
+
+The function requires at least 3 arguments:
+ - the name of the Composer package that is triggering the deprecation
+ - the version of the package that introduced the deprecation
+ - the message of the deprecation
+ - more arguments can be provided: they will be inserted in the message using `printf()` formatting
+
+Example:
+```php
```
Comment (cedriclombardot) [inline src/Symfony/Contracts/Deprecation/README.md]: 8.9 should be a string ?

Comment (greg0ire) [reply src/Symfony/Contracts/Deprecation/README.md]: See https://github.com/symfony/symfony/pull/35526#issuecomment-580464044

**Comment by derrabus** [comment]
Hey. I'm a bit late to the party here, sorry for that. I wasn't feeling well at the UG and had to leave early. Seems like I've missed an important discussion. 😃 

I really like that we're now getting a unified way to express deprecations instead of the boilerplate that we've used before. This will also make it a lot easier to make use of that mechanism in other libraries and even userland code. This is a good thing.

However, I'm not really convinced that wrapping the deprecation into an `assert()` is the way to go, mainly following @ostrolucky's arguments. To me, `zend.assertions` is a setting that is totally unrelated to deprecations. Also, I want to disable assertions on prod, but I might want to log deprecations there.

I fear that attaching this side effect to the `zend.assertions` setting encourages people to chose ill-advised production settings for their php environment. I realize that this decision has already been made. Nevertheless, I'd like to have a discussion about alternatives to the `assert()` wrapping.

I mean, all deprecations are logged with a dedicated error level (`E_USER_DEPRECATED`). Is there really no better way to not log them in production? And if that way is too complicated, what can we do to make it easier?

**Comment by derrabus** [comment]
> > In your example it looks like 8.9 is a float but it should be a string.
> 
> In PHP there is a really nice feature which is called type coercion. It's really nice because we can delegate these kind of details to the engine when they don't matter and let it deal with the concern for us!

Float to string casts are locale-dependant in php! We really should not pass around version numbers as floats. https://3v4l.org/89Wu2

**Comment by alcaeus** [comment]
The reason we decided to use `assert` is that in a production environment, it has no impact whatsoever since the parser won't generate an Opcode for the call. This is to remove concerns about `@trigger_error` having side effects and potential performance implications.

As for logging in production, I believe that we should never suggest people turn on `zend.assertions` in production, regardless for what use case without completely understanding the implications. If the only goal is to log deprecations, this should be done by creating a separate package that replaces the deprecation contract and provides its own implantation of the `deprecated` method: this method can then handle the deprecation using Monolog or any other mechanism that the user chooses to use.

**Comment by lyrixx** [comment]
> As for logging in production, I believe that we should never suggest people turn on `zend.assertions` in production, regardless for what use case without completely understanding the implications. If the only goal is to log deprecations, this should be done by creating a separate package that replaces the deprecation contract and provides its own implantation of the `deprecated` method: this method can then handle the deprecation using Monolog or any other mechanism that the user chooses to use.

Should we (symfony) provide such package? (just asking)

**Comment by derrabus** [comment]
> The reason we decided to use `assert` is that in a production environment, it has no impact whatsoever since the parser won't generate an Opcode for the call. This is to remove concerns about `@trigger_error` having side effects and potential performance implications.

Do we have any numbers on how bad the performance hit of `trigger_error` currently really is? That being said, deprecations are designed to be avoidable, meaning you should always be able to able to transform your application to a state where no deprecations are triggered at all. So even *if* there is a measurable performance penalty because of deprecations, that penalty should only be temporary.

Keep in mind that with this PR, we already have the `deprecated()` function call. That call will be compiled and executed, no matter if assertions are enabled or not. Let's say we would introduce a env variable to control this behavior:

```php
function deprecated(…)
{
    if ($_SERVER['SYMFONY_DISABLE_DEPRECATIONS'] ?? false) {
        return;
    }

    @trigger_error(…);
}
```

How much more expensive would this be, compared to the `assert()` solution with assertions disabled?

> As for logging in production, I believe that we should never suggest people turn on `zend.assertions` in production, regardless for what use case without completely understanding the implications.

My point exactly.

> If the only goal is to log deprecations, this should be done by creating a separate package that replaces the deprecation contract and provides its own implantation of the `deprecated` method:

This sounds way too complicated to me. My prediction: Unless we provide an obvious and easy way to switch this behavior, people will enable assertions in production if they want deprecations to be logged.

> this method can then handle the deprecation using Monolog or any other mechanism that the user chooses to use.

That's beside the point. Logging errors already works. Symfony's error handler does a pretty good job delegating logged errors to Monolog. And other error handler have similar features for logging errors.

**Comment by ostrolucky** [comment]
> How much more expensive would this be, compared to the assert() solution with assertions disabled?

I'm going to answer mine and your question:
```
>>>ini_set('error_reporting', E_ERROR);
=> "32767"
>>>ini_set('log_errors', 0);
=> "1"
>>>ini_set('zend.assertions', 0);
=> "1"
>>>function deprecated() { assert(@trigger_error('foo', E_USER_DEPRECATED)); }
>>>timeit -n1000 @trigger_error('foo', E_USER_DEPRECATED);
=> true
Command took 0.000007 seconds on average (0.000007 median; 0.007471 total) to complete.
>>>timeit -n1000 deprecated()
=> null
Command took 0.000003 seconds on average (0.000003 median; 0.002710 total) to complete.
>>>ini_set('zend.assertions', 1);
=> "0"
>>>timeit -n1000 deprecated()
=> null
Command took 0.000009 seconds on average (0.000008 median; 0.009156 total) to complete.
>>>
```

As for me, I'm most worried about side effects. Enabling assertions (which will now be encouraged in dev env) causes assertions in vendor/package{0..n} to execute extra code, not just vendor/packageWithDeprecations. And those packages don't expect most users having deprecations enabled, because symfony suddenly decided to hog zend.assertions for deprecations. Difficulty with logging deprecations while having disabled assertions also belongs to this category, because I agree people will likely jump into just enabling assertions in production, which will cause these side effects there as well.

**Comment by nicolas-grekas** [comment]
Please have a look at #35648
